compiler/src/scala/quoted/internal/impl/QuoteContextImpl.scala

```diff
@@ -2200,22 +2201,35 @@ class QuoteContextImpl private (ctx: Context) extends QuoteContext, QuoteUnpickl
           case _ => None
     end AmbiguousImplicitsTypeTest
 
+    type Owner = dotc.core.Symbols.Symbol
+
+    object Owner extends OwnerModule:
+      def apply(sym: Symbol): Owner =
+        assert(self.exists, "Cannot convert NoSymbol into an Owner")
+        sym
+      def spliceOwner: Owner = ctx.owner
+    end Owner
+
+    object OwnerMethodsImpl extends OwnerMethods:
+      extension (self: Owner):
+        def symbol: Symbol = self
+    end OwnerMethodsImpl
+
     type Symbol = dotc.core.Symbols.Symbol
 
     object Symbol extends SymbolModule:
-      def currentOwner(using ctx: Context): Symbol = ctx.owner
       def requiredPackage(path: String): Symbol = dotc.core.Symbols.requiredPackage(path)
       def requiredClass(path: String): Symbol = dotc.core.Symbols.requiredClass(path)
       def requiredModule(path: String): Symbol = dotc.core.Symbols.requiredModule(path)
       def requiredMethod(path: String): Symbol = dotc.core.Symbols.requiredMethod(path)
       def classSymbol(fullName: String): Symbol = dotc.core.Symbols.requiredClass(fullName)
-      def newMethod(owner: Symbol, name: String, tpe: TypeRepr): Symbol =
-        newMethod(owner, name, tpe, Flags.EmptyFlags, noSymbol)
-      def newMethod(owner: Symbol, name: String, tpe: TypeRepr, flags: Flags, privateWithin: Symbol): Symbol =
+      def newMethod(name: String, tpe: TypeRepr)(using owner: Owner): Symbol =
+        newMethod(name, tpe, Flags.EmptyFlags, noSymbol)
+      def newMethod(name: String, tpe: TypeRepr, flags: Flags, privateWithin: Symbol)(using owner: Owner): Symbol =
         dotc.core.Symbols.newSymbol(owner, name.toTermName, flags | dotc.core.Flags.Method, tpe, privateWithin)
-      def newVal(owner: Symbol, name: String, tpe: TypeRepr, flags: Flags, privateWithin: Symbol): Symbol =
+      def newVal(name: String, tpe: TypeRepr, flags: Flags, privateWithin: Symbol)(using owner: Owner): Symbol =
         dotc.core.Symbols.newSymbol(owner, name.toTermName, flags, tpe, privateWithin)
-      def newBind(owner: Symbol, name: String, flags: Flags, tpe: TypeRepr): Symbol =
+      def newBind(name: String, flags: Flags, tpe: TypeRepr)(using owner: Owner): Symbol =
         dotc.core.Symbols.newSymbol(owner, name.toTermName, flags | Case, tpe)
       def noSymbol: Symbol = dotc.core.Symbols.NoSymbol
```

Comment:
Since creating a symbol is untrivial, it might be better to be explicit about the owner instead of being implicit. This way, they have a consistent model about how symbols work, the invariants, how to debug and fix owner-related problems.

Reply:
The early owner check has proven to make simplify any bugs in the owners at the point where it is mishandled.

Reply:
I think part of the ergonomics of the API is a simple and consistent programming model. As we cannot completely hide the technical detail of owners and the owner-invariant is error-prone, it might be better to make them _explicit_:

- meta-programmers will learn a consistent model for symbols, where each symbol has an owner.
- they will be more careful when using the APIs. If they mess up with owners, they will not blame the macro system. If the argument is implicit, part of the responsibility of the owner-errors is due to the macro system.

Reply:
Then I will try to make a version where we handle owners explicitly. I believe this will not be as simple to handle.

Reply:
I added a version of the `ValDef.let` fix in #10394

Reply:
Also added #10406 with a version where owners are handled explicitly